### PR TITLE
Migration tool ObjectMetadata transforms

### PR DESCRIPTION
**Pull request by davidh44**
## Motivation and Context

Add transforms for `ObjectMetadata` setters, moving them to `PutObjectRequest`, and pass `ContentLength` to `RequestBody`

```
ContentLength
ContentEncoding
ContentType
ContentLanguage
CacheControl
ContentDisposition
ContentMD5
SSEAlgorithm
ServerSideEncryption
SSECustomerKeyMd5
BucketKeyEnabled
HttpExpiresDate
```

TODO in follow up PRs:
- Remove `ObjectMetadata` instantiation (transformed to `HeadObjectRespoonse`)

## Modifications

Updated `JavaIsoVisitor` to `JavaVisitor` in `S3PutObjectRequestToV2` in order to return a statement:

from:
`request.setMetadata(metadata);`

to:
`request = request.toBuilder().setXX(..).setYY(..).build();`

## Testing

added end to end tests

### CODE REVIEW

File: test/v2-migration-tests/src/test/resources/software/amazon/awssdk/v2migrationtests/maven-nocompile/after/src/main/java/foo/bar/S3Transforms.java [hunk 1]
```diff
@@ -15,20 +15,57 @@
...
+        /*AWS SDK for Java v2 migration: Transform for ObjectMetadata setter - httpExpiresDate - is not supported, please manually migrate the code by setting it on the v2 PutObjectRequest.*/metadata.httpExpiresDate(dateVal);
+        /*AWS SDK for Java v2 migration: Transform for ObjectMetadata setter - expirationTime - is not supported, please manually migrate the code by setting it on the v2 PutObjectRequest.*/metadata.expirationTime(dateVal);
+        /*AWS SDK for Java v2 migration: Transform for ObjectMetadata setter - restoreExpirationTime - is not supported, please manually migrate the code by setting it on the v2 PutObjectRequest.*/metadata.restoreExpirationTime(dateVal);
+
+        /*AWS SDK for Java v2 migration: Transform for ObjectMetadata setter - header - is not supported, please manually migrate the code by setting it on the v2 PutObjectRequest.*/metadata.header("key", "val");
+        /*AWS SDK for Java v2 migration: Transform for ObjectMetadata setter - addUserMetadata - is not supported, please manually migrate the code by setting it on the v2 PutObjectRequest.*/metadata.addUserMetadata("a", "b");
```
Comment (dagnir) [inline test/v2-migration-tests/src/test/resources/software/amazon/awssdk/v2migrationtests/maven-nocompile/after/src/main/java/foo/bar/S3Transforms.java:69]: This is very hard to read, can we put the comment on its own line above the line of code it's describing?

Comment (davidh44) [reply test/v2-migration-tests/src/test/resources/software/amazon/awssdk/v2migrationtests/maven-nocompile/after/src/main/java/foo/bar/S3Transforms.java:69]: I tried that, but then we can't maintain the spacing of the code itself, it would begin on the next line with no indentation. We are doing the same for other comments as well. wdyt?

File: test/v2-migration-tests/src/test/resources/software/amazon/awssdk/v2migrationtests/maven/before/src/main/java/foo/bar/S3Streaming.java [hunk 2]
```diff
@@ -90,4 +106,38 @@ void putObjectRequesterPaysSetter() {
...
+        ObjectMetadata metadata = new ObjectMetadata();
+        metadata.setContentLength(66);
+        metadata.setContentType("text/plain");
+        metadata.setContentEncoding("UTF-8");
+
+        PutObjectRequest request = new PutObjectRequest("bucket", "key", "location");
+        request.setMetadata(metadata);
+    }
+
+    void putObjectRequest_withMetadata() {
```
Comment (zoewangg) [inline test/v2-migration-tests/src/test/resources/software/amazon/awssdk/v2migrationtests/maven/before/src/main/java/foo/bar/S3Streaming.java:120]: Can we add a test with empty metadata, i.e., just `ObjectMetadata metadata = new ObjectMetadata();`

File: test/v2-migration-tests/src/test/resources/software/amazon/awssdk/v2migrationtests/maven/after/src/main/java/foo/bar/S3Streaming.java [hunk 3]
```diff
@@ -76,10 +92,12 @@ void putObject_requestPojoWithInputStream(String bucket, String key) {
 
         PutObjectRequest request1 = PutObjectRequest.builder().bucket(bucket).key(key).websiteRedirectLocation("location")
             .build();
-        /*AWS SDK for Java v2 migration: When using InputStream to upload with S3Client, Content-Length should be specified and used with RequestBody.fromInputStream(). Otherwise, the entire stream will be buffered in memory.*/s3.putObject(request1, RequestBody.fromContentProvider(() -> inputStream1, "binary/octet-stream"));
+        /*AWS SDK for Java v2 migration: When using InputStream to upload with S3Client, Content-Length should be specified and used with RequestBody.fromInputStream(). Otherwise, the entire stream will be buffered in memory.*/s3.putObject(request1, RequestBody.fromContentProvider(() -> inputStream1, "application/octet-stream"));
```
[the remark this answers is not in the document]
Comment (davidh44) [reply test/v2-migration-tests/src/test/resources/software/amazon/awssdk/v2migrationtests/maven/after/src/main/java/foo/bar/S3Streaming.java:95]: Binary was incorrect. We default to application for both v1 and v2

File: v2-migration/src/main/java/software/amazon/awssdk/v2migration/S3PutObjectRequestToV2.java [hunk 4]
```diff
@@ -235,35 +281,173 @@ private J.MethodInvocation addFileToPutObject(J.MethodInvocation method, Express
...
+
+            if (map == null) {
+                map = new HashMap<>();
+            }
 
-            return -1;
+            map.put(methodName, value);
+            metadataMap.put(variableName, map);
+
+            return null;
```
[the remark this answers is not in the document]
Comment (davidh44) [reply v2-migration/src/main/java/software/amazon/awssdk/v2migration/S3PutObjectRequestToV2.java:447]: This removes the entire line which is what we want. Adding comment to clarify

File: test/v2-migration-tests/src/test/resources/software/amazon/awssdk/v2migrationtests/maven-nocompile/after/src/main/java/foo/bar/S3Transforms.java (hunk 1 above)
Comment (davidh44) [reply test/v2-migration-tests/src/test/resources/software/amazon/awssdk/v2migrationtests/maven-nocompile/after/src/main/java/foo/bar/S3Transforms.java]: Will add new line suffix to comment and return method with maybeAutoFormat

Comment (davidh44) [reply test/v2-migration-tests/src/test/resources/software/amazon/awssdk/v2migrationtests/maven-nocompile/after/src/main/java/foo/bar/S3Transforms.java]: Nvm, that still results in code having no indentation on next line. Will keep as is

**Review by zoewangg — APPROVED** [review]
